### PR TITLE
page_service: don't count time spent flushing towards smgr latency metrics

### PR DESCRIPTION
**Pull request by problame**
## Problem

In #9962 I changed the smgr metrics to include time spent on flush.

It isn't under our (=storage team's) control how long that flush takes because the client can stop reading requests.

## Summary of changes

Stop the timer as soon as we've buffered up the response in the `pgb_writer`.

Track flush time in a separate metric.

### CODE REVIEW

File: pageserver/src/metrics.rs [hunk 1]
```diff
@@ -1272,6 +1292,50 @@ impl Drop for SmgrOpTimer {
...
+
+        now
+    }
+}
+
+impl Drop for SmgrOpTimer {
+    fn drop(&mut self) {
+        self.smgr_op_end();
+    }
+}
```
Comment (problame) [inline pageserver/src/metrics.rs:1304]: Hmm, does this double-count?

Comment (problame) [reply pageserver/src/metrics.rs:1304]: Yes it is, fixing.

Comment (problame) [reply pageserver/src/metrics.rs]: 4e4b53bf471eb46011b8b2a574c0f73b4cebe218

File: pageserver/src/metrics.rs [hunk 2]
```diff
@@ -1223,29 +1223,41 @@ pub(crate) mod virtual_file_io_engine {
...
         };
         self.throttled += *throttle;
     }
-}
 
-impl Drop for SmgrOpTimer {
-    fn drop(&mut self) {
-        let elapsed = self.start.elapsed();
+    /// Must only call once!
+    fn smgr_op_end(&self) -> Instant {
```
Comment (VladLazar) [inline pageserver/src/metrics.rs:1258]: nit: can we assert on this with an `ended` member?

File: pageserver/src/metrics.rs [hunk 3]
```diff
@@ -1272,6 +1284,85 @@ impl Drop for SmgrOpTimer {
         if let Some(per_timeline_getpage_histo) = &self.per_timeline_latency_histo {
             per_timeline_getpage_histo.observe(elapsed);
         }
+
+        now
```
Comment (VladLazar) [inline pageserver/src/metrics.rs:1288]: nit: this includes the metric update time, but probably irrelevant in practice

File: pageserver/src/metrics.rs [hunk 4]
```diff
@@ -1272,6 +1284,85 @@ impl Drop for SmgrOpTimer {
...
+                self.global_micros
+                    .inc_by(u64::try_from(elapsed.as_micros()).unwrap());
+                self.per_timeline_micros
+                    .inc_by(u64::try_from(elapsed.as_micros()).unwrap());
+                self.base = now;
+            },
+            |mut observe| {
+                observe();
+            },
+        );
```
Comment (VladLazar) [inline pageserver/src/metrics.rs:1356]: nit: this is a bit too smart for my taste. You could've done it with a select, but the comment helps.

**Review by VladLazar — APPROVED** [review]

File: pageserver/src/metrics.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/metrics.rs:1288]: The idea was to avoid a gettimeofday call

File: pageserver/src/metrics.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/metrics.rs]: Addressed in recent pushes, please check

Comment (VladLazar) [reply pageserver/src/metrics.rs]: looks good - thanks for tweaking

**Pull request merged**